libsplinter/src/rest_api/auth/identity/oauth.rs

```diff
@@ -110,10 +110,55 @@ impl IdentityProvider for OAuthUserIdentityProvider {
                     Ok(Some(user_id))
                 }
                 Ok(None) => {
-                    self.oauth_user_session_store
-                        .remove_session(token)
-                        .map_err(|err| InternalError::from_source(err.into()))?;
-                    Ok(None)
+                    // The access token didn't work; see if there's a refresh token that can be used
+                    // to get a new one.
+                    match session.oauth_refresh_token() {
+                        Some(refresh_token) => {
+                            // Try using the session's OAuth refresh token to get a new OAuth
+                            // access token
+                            match self
+                                .oauth_client
+                                .exchange_refresh_token(refresh_token.to_string())
+                            {
+                                Ok(access_token) => {
+                                    // Update the access token in the store
+                                    let updated_session = session
+                                        .into_update_builder()
+                                        .with_oauth_access_token(access_token.clone())
+                                        .build();
+                                    self.oauth_user_session_store
+                                        .update_session(updated_session)
+                                        .map_err(|err| InternalError::from_source(err.into()))?;
+                                    // Authenticate with the new access token; if this fails (we
+                                    // get Ok(None) or Err(_)), something's wrong that can't be
+                                    // handled here.
+                                    match self.oauth_client.get_subject(&access_token)? {
+                                        Some(_) => Ok(Some(user_id)),
+                                        None => Err(InternalError::with_message(
+                                            "failed to authenticate user with new access token"
+                                                .into(),
+                                        )),
+                                    }
+                                }
+                                Err(_) => {
```

Comment:
It doesn't matter what error this is? Should we be logging here?

Reply:
Changed this to log the error, but only with `debug` since this may not be a true error; it could just be that the token has been revoked. It's not easy to tell from the error provided by the oauth crate what the true cause of the failure is.